File: PrismTests/TranslationActionPolicyTests.swift
import CoreGraphics
import XCTest

@testable import Prism

final class TranslationActionPolicyTests: XCTestCase {

    private func capture(
        isEditable: Bool,
        didSelectAll: Bool = false,
        usedClipboardForRead: Bool = false
    ) -> FocusedTextCapture {
        FocusedTextCapture(
            text: "hello",
            didSelectAll: didSelectAll,
            usedClipboardForRead: usedClipboardForRead,
            isEditable: isEditable
        )
    }

    // MARK: - ⌃⌥T (.replaceInPlace)

    func testTranslateOnlyReadOnlyShowsPanel() {
        let result = TranslationActionPolicy.resolve(
            presentation: .replaceInPlace(sendAfter: false),
            capture: capture(isEditable: false, usedClipboardForRead: true)
        )
        XCTAssertTrue(result.showPanel)
        XCTAssertFalse(result.canReplace)
        XCTAssertFalse(result.sendAfter)
    }

    func testTranslateOnlyEditableFieldReplacesInPlace() {
        let result = TranslationActionPolicy.resolve(
            presentation: .replaceInPlace(sendAfter: false),
            capture: capture(isEditable: true)
        )
        XCTAssertFalse(result.showPanel)
        XCTAssertFalse(result.canReplace)
    }

    func testTranslateOnlyComposeClipboardReplacesInPlace() {
        // Simulates readFallbackFailed clipboard path (Discord compose with AX focus).
        let result = TranslationActionPolicy.resolve(
            presentation: .replaceInPlace(sendAfter: false),
            capture: capture(isEditable: true, usedClipboardForRead: true)
        )
        XCTAssertFalse(result.showPanel)
        XCTAssertFalse(result.canReplace)
    }

    func testTranslateAndSendSetsSendAfter() {
        let result = TranslationActionPolicy.resolve(
            presentation: .replaceInPlace(sendAfter: true),
            capture: capture(isEditable: true)
        )
        XCTAssertFalse(result.showPanel)
        XCTAssertTrue(result.sendAfter)
    }

    func testSelectAllReadOnlyShowsPanel() {
        let result = TranslationActionPolicy.resolve(
            presentation: .replaceInPlace(sendAfter: false),
            capture: capture(isEditable: false, didSelectAll: true, usedClipboardForRead: true)
        )
        XCTAssertTrue(result.showPanel)
    }

    // MARK: - ⌃⌥Y (.popup)

    func testPopupAlwaysShowsPanel() {
        let result = TranslationActionPolicy.resolve(
            presentation: .popup,
            capture: capture(isEditable: false)
        )
        XCTAssertTrue(result.showPanel)
        XCTAssertFalse(result.canReplace)
        XCTAssertFalse(result.sendAfter)
    }

    func testPopupCanReplaceWhenEditable() {
        let result = TranslationActionPolicy.resolve(
            presentation: .popup,
            capture: capture(isEditable: true)
        )
        XCTAssertTrue(result.showPanel)
        XCTAssertTrue(result.canReplace)
    }

    func testPopupReadOnlyStillCopyOnly() {
        let result = TranslationActionPolicy.resolve(
            presentation: .popup,
            capture: capture(isEditable: false, usedClipboardForRead: true)
        )
        XCTAssertTrue(result.showPanel)
        XCTAssertFalse(result.canReplace)
    }

    // MARK: - shouldReplaceInField

    func testShouldReplaceInFieldFollowsCaptureEditableOnly() {
        XCTAssertFalse(
            TranslationActionPolicy.shouldReplaceInField(
                preferReplaceInPlace: true,
                capture: capture(isEditable: false, usedClipboardForRead: true)
            )
        )
        XCTAssertTrue(
            TranslationActionPolicy.shouldReplaceInField(
                preferReplaceInPlace: true,
                capture: capture(isEditable: true)
            )
        )
    }

    // MARK: - Clipboard editable flag (AX context, no mouse)

    func testReadFallbackFailedIsEditableForClipboard() {
        XCTAssertTrue(
            FocusedTextIO.editableFlagForClipboardCapture(
                preferReplaceInPlace: true,
                axError: .readFallbackFailed,
                isEditable: true
            )
        )
    }

    func testNoFocusedElementClipboardIsNeverEditable() {
        XCTAssertFalse(
            FocusedTextIO.editableFlagForClipboardCapture(
                preferReplaceInPlace: true,
                axError: .noFocusedElement,
                isEditable: false
            )
        )
    }

    func testNoSelectionInReadOnlyClipboardIsNeverEditable() {
        XCTAssertFalse(
            FocusedTextIO.editableFlagForClipboardCapture(
                preferReplaceInPlace: true,
                axError: .noSelectionInReadOnly,
                isEditable: false
            )
        )
    }
}

File: Prism.Windows/tests/Prism.Core.Tests/TranslationActionPolicyTests.cs
using Prism.Core.Policy;
using Prism.Core.TextIO;

namespace Prism.Core.Tests;

public class TranslationActionPolicyTests
{
    private static FocusedTextCapture Capture(
        bool isEditable,
        bool didSelectAll = false,
        bool usedClipboardForRead = false) =>
        new("hello", didSelectAll, usedClipboardForRead, isEditable);

    [Fact]
    public void TranslateOnly_ReadOnly_ShowsPanel()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.ReplaceInPlace(sendAfter: false),
            Capture(isEditable: false, usedClipboardForRead: true));

        Assert.True(result.ShowPanel);
        Assert.False(result.CanReplace);
        Assert.False(result.SendAfter);
    }

    [Fact]
    public void TranslateOnly_EditableField_ReplacesInPlace()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.ReplaceInPlace(sendAfter: false),
            Capture(isEditable: true));

        Assert.False(result.ShowPanel);
        Assert.False(result.CanReplace);
    }

    [Fact]
    public void TranslateOnly_ComposeClipboard_ReplacesInPlace()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.ReplaceInPlace(sendAfter: false),
            Capture(isEditable: true, usedClipboardForRead: true));

        Assert.False(result.ShowPanel);
        Assert.False(result.CanReplace);
    }

    [Fact]
    public void TranslateAndSend_SetsSendAfter()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.ReplaceInPlace(sendAfter: true),
            Capture(isEditable: true));

        Assert.False(result.ShowPanel);
        Assert.True(result.SendAfter);
    }

    [Fact]
    public void SelectAll_ReadOnly_ShowsPanel()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.ReplaceInPlace(sendAfter: false),
            Capture(isEditable: false, didSelectAll: true, usedClipboardForRead: true));

        Assert.True(result.ShowPanel);
    }

    [Fact]
    public void Popup_AlwaysShowsPanel()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.Popup(),
            Capture(isEditable: false));

        Assert.True(result.ShowPanel);
        Assert.False(result.CanReplace);
        Assert.False(result.SendAfter);
    }

    [Fact]
    public void Popup_CanReplace_WhenEditable()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.Popup(),
            Capture(isEditable: true));

        Assert.True(result.ShowPanel);
        Assert.True(result.CanReplace);
    }

    [Fact]
    public void Popup_ReadOnly_StillCopyOnly()
    {
        var result = TranslationActionPolicy.Resolve(
            TranslationPresentation.Popup(),
            Capture(isEditable: false, usedClipboardForRead: true));

        Assert.True(result.ShowPanel);
        Assert.False(result.CanReplace);
    }

    [Fact]
    public void ShouldReplaceInField_FollowsCaptureEditableOnly()
    {
        Assert.False(TranslationActionPolicy.ShouldReplaceInField(
            preferReplaceInPlace: true,
            Capture(isEditable: false, usedClipboardForRead: true)));
        Assert.True(TranslationActionPolicy.ShouldReplaceInField(
            preferReplaceInPlace: true,
            Capture(isEditable: true)));
    }

    [Fact]
    public void ReadFallbackFailed_IsEditableForClipboard()
    {
        Assert.True(FocusedTextCaptureRules.EditableFlagForClipboardCapture(
            preferReplaceInPlace: true,
            FocusedTextIoErrorKind.ReadFallbackFailed,
            isEditable: true));
    }

    [Fact]
    public void NoFocusedElement_Clipboard_IsNeverEditable()
    {
        Assert.False(FocusedTextCaptureRules.EditableFlagForClipboardCapture(
            preferReplaceInPlace: true,
            FocusedTextIoErrorKind.NoFocusedElement,
            isEditable: false));
    }

    [Fact]
    public void NoSelectionInReadOnly_Clipboard_IsNeverEditable()
    {
        Assert.False(FocusedTextCaptureRules.EditableFlagForClipboardCapture(
            preferReplaceInPlace: true,
            FocusedTextIoErrorKind.NoSelectionInReadOnly,
            isEditable: false));
    }
}
